extension/js/common/core/crypto/pgp/openpgp-key.ts

```diff
@@ -192,13 +192,37 @@ export class OpenPGPKey {
     }
     const algoInfo = opgpKey.primaryKey.getAlgorithmInfo();
     const key = keyToUpdate || {} as Key; // if no key to update, use empty object, will get props assigned below
+    const encryptionKey = await Catch.undefinedOnException(opgpKey.getEncryptionKey());
+    const encryptionKeyIgnoringExpiration = encryptionKey ? encryptionKey : await OpenPGPKey.getExpiredKey(opgpKey.getEncryptionKey.bind(opgpKey), exp, expired);
+    const signingKey = await Catch.undefinedOnException(opgpKey.getSigningKey());
+    const signingKeyIgnoringExpiration = signingKey ? signingKey : await OpenPGPKey.getExpiredKey(opgpKey.getSigningKey.bind(opgpKey), exp, expired);
+    const paramCountByAlgo = {
```

Comment:
Since this is a constant, it should be brought out of this method, to be a private property on this class

Reply:
I tried to, but failed to initialize it because `opgp` was undefined at the moment of constant initialization....

Reply:
Ah. Let's consider a private method - maybe `checkMissingPrivateKeyPackets`. That way you can remove the whole content of lines 199 to 216 and put it there, with a comment to explain why is `paramCountByAlgo` not a class constant. The method could return `{missingPrivateKeyForSigning, missingPrivateKeyForDecryption}`. Also, the method should have a link to this issue 2887 so that we can refer to it in the future.

Reply:
@tomholub If we take it out to a separate method, than `usableForSigning` may return `true` while it's impossible to sign (missing packets). Well, it's up to you to decide, of course. We can also try to make lazy initialization on first access to `paramCountByAlgo`, filling it if it is null

Reply:
I leave it up to you then, you're the one working on the PR, so you'll know better.

Reply:
UPD: yes, a private method called from `convertExternalLibraryObjToKey` -- good idea, I'll place some logic there.